### PR TITLE
cherry-pick 339b531 (unet support)

### PR DESCRIPTION
**Pull request by cool-bigdogs-tshirt**
## Description

had to make these changes to get https://github.com/AUTOMATIC1111/stable-diffusion-webui-tensorrt working

## Notes

cherry-picked https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/339b5315700a469f4a9f0d5afc08ca2aca60c579

had to add the apply_optimizations method back in a hacky way. passing in `'None'` will unload the optimizations, passing in no args will re-apply them, which seems to be the intent of the extension.

## Environment and Testing

tested on arch:
```
22:50:35-916899 INFO     Torch 2.0.1
22:50:35-932185 DEBUG    Torch allocator: native
22:50:35-936831 INFO     Torch backend: nVidia CUDA 12.1 cuDNN 8900
22:50:35-953974 INFO     Torch detected GPU: NVIDIA GeForce RTX 4070 Ti VRAM 12007 Arch (8, 9) Cores 60
```
was able to convert to onxx, and it tried running the tensorrt compile, but core dumped so I haven't been able to test further.
running the `trtexec` command on its own also crashes, so i don't think it's a webui issue.

### CODE REVIEW

**Comment by cool-bigdogs-tshirt** [comment]
awesome, nice to hear! i was able to compile to trt exactly once, but inference failed so i haven't been able to test fully in the UI yet. trying to upgrade TensorRT to the GA release instead of EA now to see if that helps...

if anyone else has been running in to core dumps when compiling or running inference, let me know. i think it has something to do with running out of GPU memory, which TensorRT does not handle nicely, but i'm not sure

File: modules/sd_hijack.py [hunk 1]
```diff
@@ -154,6 +154,14 @@ class StableDiffusionModelHijack:
     def __init__(self):
         self.embedding_db.add_embedding_dir(opts.embeddings_dir)
 
+    def apply_optimizations(self, *args):
```
Comment (cool-bigdogs-tshirt) [inline modules/sd_hijack.py:157]: this is hacky, but the only mismatch between this repo and what the addon expects.

i have other issues with the addon code though, mainly with the way it tries to find TensorRT. i have it installed on my PATH already but it *really* wants it in a subdirectory...

**Comment by Ataraksia** [comment]
Yeah, I did end up just placing the TensorRT folder within the extension folder to avoid issues with that, which I don't really mind doing myself, but obviously not the best solution ever.  Also, I actually did hit a core dump as well, albeit it was when attempting to convert a SD2.1 model, which I expected to have issues with based on what I read in A1111 extension repo. 

Definitely a good idea updating to the TensorRT 8.6 GA release just to be sure that's not causing issues, that's what I'm using as well.  Also, if you're trying to convert with a maximum batch size setting above 1, try dropping it just to make sure that's not causing an issue as I've noticed the conversion process has some giant VRAM spikes when a maximum batch size above 1 is specified.

**Comment by Ataraksia** [comment]
Oh also, definitely try a fresh automatic install if you haven't already.  Even after I managed to convert a model properly, I couldn't get inferencing to run properly on my non-stock install.

**Comment by cool-bigdogs-tshirt** [comment]
yeah, i tried batch sizes 1-4 and that didn't work, but even with just batch size 1 the only time it compiled successfully was without `--fp16` which I'm guessing is why the inference is crashing... I never actually *see* the vram spike though :shrug:

**Comment by Ataraksia** [comment]
Hmm, that's interesting that it only works without --fp16.  I'm currently testing various trtexec build options because I've noticed that converting with batch sizes above 1, while it succeeds, produces some very low quality images.  I'm trying to see if there's anything in the conversion process I can tweak to solve that.   Also, when I mentioned VRAM spikes, I was referring to ones that exceeded my total VRAM entirely, and so they didn't actually show up in monitoring software, I only know about them because of the failure to allocate memory messages that popped up in my terminal.

**Comment by cool-bigdogs-tshirt** [comment]
hm... managed to get one wicked fast inference, but hires fix was on and it is not happy lol. how nicely are these compiled models supposed to play with image/batch sizes they weren't compiled for?

**Comment by cool-bigdogs-tshirt** [comment]
frustrating that it can't actually catch the oom error, i have no idea if it's close to being able to fit or not.
i can still only compile batch size 1, which is cool and fast, but not worth losing hires fix, loras, etc...

**Comment by Ataraksia** [comment]
So after some more testing on my end, it would appear the hires fix actually does work with batches still, just one less than the specified maximum, which is good enough for me.  I don't know why hires isn't working at all on your end.  

Also,  another rather interesting observation, it would appear loras are still functioning with the TensorRT model activated... sorta kinda.  I'm not sure if it's related to the changes @vladmandic has made to loras or not, but at least on my machine, while the effect seems to be lessened compared to a standard model, there is definitely still an effect.  Additionally, while there does seem to be a small speed penalty when using a lora, the performance improvement is still upwards of 50%+ compared to the standard model so it's not just automatic falling back to the standard model when a lora is applied which was my first guess.  

The 768x768 cap is unfortunate, although over on the extension page, someone apparently had some success with using tiled diffusion to upscale images, so that could potentially be an acceptable option.  

One final observation I've had is that when Negative Guidance minimum sigma is enabled, inferencing with batch size 1 fails with some samplers.  That can be resolved in a few ways;  I don't know enough about the implementation for Negative Guidance minimum sigma to be able to say if it's something that could be tweaked on that end, but it's a fairly simple change to the TensorRT extension to resolve the problem, with little to no impact as far as I can tell to the resulting TensorRT engine.  If/when vlad you decide you want to include this PR in automatic, you might also want to simply add the TensorRT extension to the built-in group.  The following tweak to that extension's export_trt.py file resolves the conflict:
```
x_min = f"{min_bs * 1}x4x{min_height // 8}x{min_width // 8}"
x_max = f"{max_bs * 2}x4x{max_height // 8}x{max_width // 8}"
context_min = f"{min_bs * 1}x{min_token_count // 75 * 77}x{cond_dim}"
context_max = f"{max_bs * 2}x{max_token_count // 75 * 77}x{cond_dim}"
timestamps_min = f"{min_bs * 1}"
timestamps_max = f"{max_bs * 2}"
```

**Comment by vladmandic** [comment]
i went through the pr and seems clean enough - anyone has any concerns in merging this? out-of-the-box concerns where this may come into play for non-tensorrt workflows?

**Comment by cool-bigdogs-tshirt** [comment]
kind of annoying that the unet is buried in settings and doesn't change with your model (from what i saw), but if you don't set it then it shouldn't affect anything else. was using it for a while last night with no unet.

**Comment by cool-bigdogs-tshirt** [comment]
also, for anyone else trying this stuff out, i highly recommend just running the tensorrt container: `docker run --gpus all -it --rm -v ./models/:/models nvcr.io/nvidia/tensorrt:23.05-py3`
hasn't segfaulted on me yet (though we'll see how it goes when inferencing).

**Comment by vladmandic** [comment]
> kind of annoying that the unet is buried in settings and doesn't change with your model (from what i saw), but if you don't set it then it shouldn't affect anything else. was using it for a while last night with no unet.

that is easily addressable once underlying code is actually fully working. i can make it behave like vae does right now, so you could set it to auto and then it would do name-matching automatically. but that's usability - lets get it working reliably first.

**Comment by cool-bigdogs-tshirt** [comment]
> i highly recommend just running the tensorrt container

lol nevermind, this crap is so annoying: `The engine plan file is not compatible with this version of TensorRT, expecting library version 8.6.1.6 got 8.6.1.2, please rebuild.`
still going to try the container, but i gotta build it from scratch i guess because the tagged version they have isn't actually the latest tensorrt release...

**Comment by vladmandic** [comment]
That'd one of my biggest issues with optimized models - they are soooo platofrm/version specific.

**Comment by Ataraksia** [comment]
> kind of annoying that the unet is buried in settings and doesn't change with your model (from what i saw), but if you don't set it then it shouldn't affect anything else. was using it for a while last night with no unet.

Hmm? Did you try setting it to auto?  Don't feel bad if it didn't jump out at you either;  I was annoyed by having to go into settings to change it each time as well,  and it took me at least half a dozen times before it finally dawned on me that the auto setting in the menu activates the corresponding TensorRT engines when you select the standard model of the same name in the list. Or did the auto setting just not work in your case?

**Comment by cool-bigdogs-tshirt** [comment]
> Did you try setting it to auto?

ah, yeah i think it was working, but i named my models with batch size info in them.

**Comment by vladmandic** [comment]
i've created a new branch and merged changes from here to that branch, it should be much easier for testing and any further work.

**Comment by vladmandic** [comment]
@cool-bigdogs-tshirt let me know if you're interested in actually bringing this to production-ready?

**Comment by cool-bigdogs-tshirt** [comment]
> @cool-bigdogs-tshirt let me know if you're interested in actually bringing this to production-ready?

never was able to get the models compiling and running well enough to actually test much unfortunately, so i haven't really messed with tensorrt since.

**Pull request merged**